### PR TITLE
The git:// protocol with proxy is not allowed

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://github.com/openshift/origin/issues/5944

### CODE REVIEW

**Comment by mfojtik** [comment]
@liggitt PTAL

File: pkg/build/api/validation/validation.go [hunk 1]
```diff
@@ -145,6 +145,10 @@ func validateBuildSpec(spec *buildapi.BuildSpec) fielderrors.ValidationErrorList
 
 const maxDockerfileLengthBytes = 60 * 1000
 
+func hasProxy(source *buildapi.GitBuildSource) bool {
+	return !(len(source.HTTPProxy) == 0 && len(source.HTTPSProxy) == 0)
```
Comment (liggitt) [inline pkg/build/api/validation/validation.go]: `len(…) > 0 || len(…) > 0` reads easier to me… what do you think?

File: pkg/build/api/validation/validation.go [hunk 2]
```diff
@@ -241,6 +245,9 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
 	if len(git.HTTPSProxy) != 0 && !isValidURL(git.HTTPSProxy) {
 		allErrs = append(allErrs, fielderrors.NewFieldInvalid("httpsproxy", git.HTTPSProxy, "proxy is not a valid url"))
 	}
+	if strings.HasPrefix(git.URI, "git://") && hasProxy(git) {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("uri", git.URI, "git:// protocol cannot be used with HTTP proxy"))
```
Comment (liggitt) [inline pkg/build/api/validation/validation.go]: HTTP or HTTPS

Comment (liggitt) [inline pkg/build/api/validation/validation.go]: ssh as well, right? Shouldn't we parse the source uri and require http or https scheme?

Comment (mfojtik) [inline pkg/build/api/validation/validation.go]: @liggitt good catch, yeah if the http* proxy is up, we should allow only http(s)://

**Comment by mfojtik** [comment]
@liggitt fixed, thanks!

File: pkg/build/api/validation/validation.go [hunk 3]
```diff
@@ -241,6 +245,9 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
 	if len(git.HTTPSProxy) != 0 && !isValidURL(git.HTTPSProxy) {
 		allErrs = append(allErrs, fielderrors.NewFieldInvalid("httpsproxy", git.HTTPSProxy, "proxy is not a valid url"))
 	}
+	if hasProxy(git) && !strings.HasPrefix(git.URI, "http") {
```
Comment (liggitt) [inline pkg/build/api/validation/validation.go]: parse the URL and check the scheme... I always clone ssh repos as `httpuser@myrepo.com` :)

Comment (mfojtik) [inline pkg/build/api/validation/validation.go]: you're weird Jordan :-) sure thing, I will fix this.

**Comment by mfojtik** [comment]
@liggitt good catch, added url parser

File: pkg/build/api/validation/validation_test.go [hunk 4]
```diff
@@ -212,6 +212,45 @@ func TestValidateBuildUpdate(t *testing.T) {
...
+					To: &kapi.ObjectReference{
+						Kind: "DockerImage",
+						Name: "repository/data",
+					},
+				},
+			},
+		},
+	}
+	errors := ValidateBuildConfig(buildConfig)
+	if len(errors) > 1 {
```
Comment (liggitt) [inline pkg/build/api/validation/validation_test.go]: `!= 1`

**Comment by liggitt** [comment]
nit on test, LGTM otherwise

**Comment by mfojtik** [comment]
@liggitt fixed, thanks [merge]

**Comment by mfojtik** [comment]
[merge]

**Comment by mfojtik** [comment]
[merge] !

**Comment by mfojtik** [comment]
[merge]

**Pull request merged**